deso/index.go

```diff
@@ -24,171 +24,238 @@ const (
-		})
-	})
-	if err != nil {
-		return nil, err
-	}
-
-	return opBundle.UtxoOpBundle, nil
+	return lib.GetUtxoOperationsForBlock(index.chainDB, index.snapshot, blockHash)
 }
 
 //
 // Balance Snapshots
 //
 
-func balanceSnapshotKey(isLockedBalance bool, publicKey *lib.PublicKey, blockHeight uint64, balance uint64) []byte {
-	startPrefix := PrefixBalanceSnapshots
-	if isLockedBalance {
-		startPrefix = PrefixLockedBalanceSnapshots
+type BalanceType uint8
+
+const (
+	DESOBalance                BalanceType = 0
+	CreatorCoinLockedBalance   BalanceType = 1
+	ValidatorStakedDESOBalance BalanceType = 2
+	LockedStakeDESOBalance     BalanceType = 3
+)
+
+func (balanceType BalanceType) BalanceTypePrefix() byte {
+	switch balanceType {
+	case DESOBalance:
+		return PrefixBalanceSnapshots
+	case CreatorCoinLockedBalance:
+		return PrefixLockedBalanceSnapshots
+	case ValidatorStakedDESOBalance:
+		return PrefixValidatorStakedBalanceSnapshots
+	case LockedStakeDESOBalance:
+		return PrefixLockedStakeBalanceSnapshots
+	default:
+		panic("unknown balance type")
 	}
+}
 
-	prefix := append([]byte{}, startPrefix)
+func balanceSnapshotKey(balanceType BalanceType, publicKey *lib.PublicKey, blockHeight uint64, balance uint64) []byte {
+	prefix := append([]byte{}, balanceType.BalanceTypePrefix())
 	prefix = append(prefix, publicKey[:]...)
 	prefix = append(prefix, lib.EncodeUint64(blockHeight)...)
 	prefix = append(prefix, lib.EncodeUint64(balance)...)
 	return prefix
 }
 
-func hypersyncHeightToBlockKey(blockHeight uint64, isLocked bool) []byte {
+func lockedStakeBalanceSnapshotKey(
+	balanceType BalanceType, stakerPKID *lib.PKID, validatorPKID *lib.PKID, lockedAtEpochNumber uint64,
+	blockHeight uint64, balance uint64,
+) []byte {
+	prefix := append([]byte{}, balanceType.BalanceTypePrefix())
+	prefix = append(prefix, stakerPKID[:]...)
+	prefix = append(prefix, validatorPKID[:]...)
+	prefix = append(prefix, lib.EncodeUint64(lockedAtEpochNumber)...)
+	prefix = append(prefix, lib.EncodeUint64(blockHeight)...)
+	prefix = append(prefix, lib.EncodeUint64(balance)...)
+	return prefix
+}
 
-	lockedByte := byte(0)
-	if isLocked {
-		lockedByte = byte(1)
+func (balanceType BalanceType) HypersyncBalanceTypePrefix() byte {
+	switch balanceType {
+	case DESOBalance:
+		return 0
+	case CreatorCoinLockedBalance:
+		return 1
+	case ValidatorStakedDESOBalance:
+		return 2
+	case LockedStakeDESOBalance:
+		return 3
+	default:
+		panic("unknown balance type")
 	}
+}
 
+func hypersyncHeightToBlockKey(blockHeight uint64, balanceType BalanceType) []byte {
 	prefix := append([]byte{}, PrefixHypersyncBlockHeightToBalances)
 	prefix = append(prefix, lib.EncodeUint64(blockHeight)...)
-	prefix = append(prefix, lockedByte)
+	prefix = append(prefix, balanceType.HypersyncBalanceTypePrefix())
 	return prefix
 }
 
-func (index *RosettaIndex) PutHypersyncBlockBalances(blockHeight uint64, isLocked bool, balances map[lib.PublicKey]uint64) {
-
+func (index *RosettaIndex) PutHypersyncBlockBalances(
```

Comment:
As discussed, you'll need to change these Puts to take a WriteBatch or whatever but I'll assume we'll handle this in a follow-up PR.